README.md
Python Linkedin API
======

[![Travis CI](https://travis-ci.org/johnidm/PyLinkedinAPI.svg)](https://travis-ci.org/johnidm/PyLinkedinAPI)
[![PyPI](https://img.shields.io/pypi/v/PyLinkedinAPI.svg)](https://pypi.python.org/pypi/PyLinkedinAPI)
[![Coverage Status](https://coveralls.io/repos/johnidm/PyLinkedinAPI/badge.svg?branch=master&service=github)](https://coveralls.io/github/johnidm/PyLinkedinAPI?branch=master)

The Python Linkedin provides a easy interface to the Linkedin API.

We are in development, if you want to help, please see [CONTRIBUTING](https://github.com/johnidm/PyLinkedinAPI/blob/master/CONTRIBUTING.rst)

### Installation

You can to install through pip.

```python
pip install PyLinkedinAPI
```

### Generate Access Token

If you don't have a access token I recommend to use the [requests-oauthlib](https://github.com/requests/requests-oauthlib)

### Examples

See examples [here](https://github.com/johnidm/PyLinkedinAPI/blob/master/examples/basic.py)

You need to generate temporary access token for basic tests:

* Acces the https://developer.linkedin.com/rest-console
* On then Authentication menu select OAuth2
* After you need to login and authorization to access some information from your LinkedIn profile
* Send anywhere request URL, for example https://api.linkedin.com/v1/people/~?format=json, and copy field access token

Run `python examples/basic.py` and insert the access token

### Created an instance

```python
from PyLinkedinAPI.PyLinkedinAPI import PyLinkedinAPI

access_token = 'AQVaE34Qblm6uIhh3wVLXuhQPSI...'

linkedin = PyLinkedinAPI(access_token)
```

#### Features
--------

Get data profile

```python
profile = linkedin.get_basic_profile()

print(profile)
```

Get data companies

```python
companies = linkedin.get_companies()

print(companies)
```

Publish comment on profile

```python
linkedin.publish_profile_comment('This is my first package in Python')
```

Publish comment on company


```python
linkedin.publish_company_comment(5470551, 'This is my first package in Python')
```

Publush Content/Comment on profile

```python
linkedin.publish_profile(comment,
                         title=title, description=description,
                         submitted_url='http://www.johnidouglas.com.br/',
                         submitted_image_url='http://www.johnidouglas.com.br/logo.jpeg')
```

Publush Content/Comment on company

```python
linkedin.publish_company(5470551,
                         comment,
                         title=title, description=description,
                         submitted_url='http://www.johnidouglas.com.br/',
                         submitted_image_url='http://www.johnidouglas.com.br/logo.jpeg)

```


### Utils
--------

* REST Console

	https://developer.linkedin.com/rest-console







HISTORY.rst
.. :changelog:

History
-------

0.1.6 (2016-01-02)
---------------------

* Improvements on project structure
* Improvements on documentation
* Added continuous delivery with Travis
* Added Python support to version 2.7.x and 3.4.x

0.1.5 (2015-10-10)
---------------------

* Documentation improvement
* Add method get_profile - fields of profile: https://developer.linkedin.com/docs/fields

0.1.4 (2015-10-09)
---------------------

* Added support to share only comment field.

0.1.3 (2015-10-09)
---------------------

* Fixed bugs.

0.1.2 (2015-05-09)
---------------------

* Fixed bugs.

0.1.1 (2015-05-09)
---------------------

* Added support to share on LinkedIn a collection of content fields.

0.1.0 (2015-05-09)
---------------------

* First release on PyPI.

tests/test_linkedin_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import unittest
import httpretty
import json

from sure import expect

from PyLinkedinAPI.PyLinkedinAPI import PyLinkedinAPI
from PyLinkedinAPI.PyLinkedinAPI import PyLinkedinAPIClientError


class TestPyLinkedin(unittest.TestCase):

    URL_GET_BASIC_PROFILE = "https://api.linkedin.com/v1/people/~?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json"

    URL_GET_COMPANIES = "https://api.linkedin.com/v1/companies:(id,name,logo-url,company-type)?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json&is-company-admin=true"

    GET_PROFILE_ERROR_403 = "https://api.linkedin.com/v1/people/~?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json"

    GET_PROFILE_FIELDS = 'https://api.linkedin.com/v1/people/~:(id,num-connections,picture-url,email-address)?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json'

    GET_PROFILE_FIELD_NOT_FOUND = 'https://api.linkedin.com/v1/people/~:(id,num-connections,data-profile)?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json'

    POST_COMMENT_PROFILE = 'https://api.linkedin.com/v1/people/~/shares?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json'

    POST_COMMENT_COMPANY = 'https://api.linkedin.com/v1/companies/536454/shares?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json'

    POST_COMMENT_COMPANY_ERROR_403 = POST_COMMENT_COMPANY

    def setUp(self):
        access_token = 'AQVaE34QblVhvPlUn-6zWh3YLgHjx...'
        self.linkedin = PyLinkedinAPI(access_token)

    @httpretty.activate
    def test_get_basic_profile(self):

        data = '''
            {
                "id": "RzvdfgdfgGT",
                "firstName": "Johni",
                "lastName": "Douglas Marangon",
                "headline": "Senior Software Engineer",
                "siteStandardProfileRequest":  {
                    "url": "https://www.linkedin.com/profile/v..."
                }
            }
        '''

        httpretty.register_uri(httpretty.GET,
                               TestPyLinkedin.URL_GET_BASIC_PROFILE,
                               body=data,
                               content_type="application/json")

        basic_profile = self.linkedin.get_basic_profile()
        data = json.loads(data)
        expect(basic_profile).to.equal(data)

    @httpretty.activate
    def test_get_companies(self):
        data = '''
            {
              "_total": 1,
              "values":  [
                 {
                  "companyType":  {
                    "code": "O",
                    "name": "Self Owned"
                  },
                  "id": 5470551,
                  "logoUrl": "https://media.licdn.com/mpr/mpr/e34...Po.png",
                  "name": "Johni Corp"
                }
              ]
            }
        '''

        httpretty.register_uri(httpretty.GET,
                               TestPyLinkedin.URL_GET_COMPANIES,
                               body=data,
                               content_type="application/json")

        basic_profile = self.linkedin.get_companies()
        data = json.loads(data)
        expect(basic_profile).to.equal(data)

    @httpretty.activate
    def test_get_profile_error_403(self):
        data = '''
        {
              "errorCode": 0,
              "message": "Member XXXX does not have permission to get company 9898",
              "requestId": "YYYY",
              "status": 403,
              "timestamp": 1441470368132
            }
        '''

        httpretty.register_uri(httpretty.GET,
                               TestPyLinkedin.GET_PROFILE_ERROR_403,
                               body=data,
                               status=403,
                               content_type="application/json")

        with self.assertRaises(PyLinkedinAPIClientError) as ex:
            self.linkedin.get_basic_profile()

        expect(str(ex.exception)).to.equal(
            'Member XXXX does not have permission to get company 9898')

    @httpretty.activate
    def test_get_profile_fields(self):
        data = '''
            {
                "id": "RdfgsgfGT",
                "numConnections": 194,
                "pictureUrl": "https://media.licdn.com/mpr/mprx/0_v0z...jQx",
                "emailAddress": "johni@johni.com"
            }
        '''
        httpretty.register_uri(httpretty.GET,
                               TestPyLinkedin.GET_PROFILE_FIELDS,
                               body=data,
                               status=200,
                               content_type="application/json")

        basic_profile = self.linkedin.get_profile(
            ['id', 'num-connections', 'picture-url', 'email-address'])
        data = json.loads(data)
        expect(basic_profile).to.equal(data)

    @httpretty.activate
    def test_get_profile_field_not_found(self):
        data = '''
            {
              "errorCode": 0,
              "message": "Unknown field {data-profile} in resource {Person}",
              "requestId": "YUUYSSII",
              "status": 400,
              "timestamp": 1444524410254
            }
        '''

        httpretty.register_uri(httpretty.GET,
                               TestPyLinkedin.GET_PROFILE_FIELD_NOT_FOUND,
                               body=data,
                               status=400,
                               content_type="application/json")

        with self.assertRaises(PyLinkedinAPIClientError) as ex:
            self.linkedin.get_profile(
                ['id', 'num-connections', 'data-profile'])

        expect(str(ex.exception)).to.equal(
            'Unknown field {data-profile} in resource {Person}')

    @httpretty.activate
    def test_publish_comment_on_profile(self):
        response_body = '''{
            "updateKey": "UPDATE-234454664-64564654646464634502",
            "updateUrl": "https://www.linkedin.com/updates?discuss=&scope=2344546642&stype=M&topic=6456465464646463450&type=U&a=1gON"
        }'''

        httpretty.register_uri(httpretty.POST,
                               TestPyLinkedin.POST_COMMENT_PROFILE,
                               body=response_body,
                               status=201,
                               content_type="application/json")

        comment = 'A laugh a day keeps the doctor away'
        response = self.linkedin.publish_profile_comment(comment)
        actual = json.loads(response_body)
        expect(actual).to.equal(response)

    @httpretty.activate
    def test_publish_comment_on_company(self):
        response_body = '''{
            "updateKey": "UPDATE-234454664-64564654646464634502",
            "updateUrl": "https://www.linkedin.com/updates?discuss=&scope=2344546642&stype=M&topic=6456465464646463450&type=U&a=1gON"
        }'''

        httpretty.register_uri(httpretty.POST,
                               TestPyLinkedin.POST_COMMENT_COMPANY,
                               body=response_body,
                               status=201,
                               content_type="application/json")

        comment = 'Have You Tried Turning It Off And On Again?'
        response = self.linkedin.publish_company_comment(536454, comment)
        actual = json.loads(response_body)
        expect(actual).to.equal(response)

    @httpretty.activate
    def test_publish_comment_on_company_error_403(self):
        response_body = '''{
              "errorCode": 0,
              "message": "Unauthorized request",
              "requestId": "2B5UXDUSL2",
              "status": 403,
              "timestamp": 1451574899170
            }'''

        httpretty.register_uri(httpretty.POST,
                               TestPyLinkedin.POST_COMMENT_COMPANY_ERROR_403,
                               body=response_body,
                               status=403,
                               content_type="application/json")

        with self.assertRaises(PyLinkedinAPIClientError) as ex:
            comment = 'Have You Tried Turning It Off And On Again?'
            self.linkedin.publish_company_comment(536454, comment)

        expect(str(ex.exception)).to.equal(
            'Unauthorized request')

    @httpretty.activate
    def test_publish_profile(self):
        pass

    @httpretty.activate
    def test_publish_company(self):
        pass


tests/test_py_linkedin_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import unittest
import httpretty
import json

from sure import expect

from PyLinkedinAPI.PyLinkedinAPI import PyLinkedinAPI
from PyLinkedinAPI.PyLinkedinAPI import PyLinkedinAPIClientError


class TestPyLinkedin(unittest.TestCase):

    def setUp(self):
        access_token = 'AQVaE34QblVhvPlUn-6zWh3YLgHjx...'
        self.linkedin = PyLinkedinAPI(access_token)

    @httpretty.activate
    def test_get_basic_profile(self):

        data = '''
            {
                "id": "RzvdfgdfgGT",
                "firstName": "Johni",
                "lastName": "Douglas Marangon",
                "headline": "Senior Software Engineer",
                "siteStandardProfileRequest":  {
                    "url": "https://www.linkedin.com/profile/v..."
                }
            }
        '''

        httpretty.register_uri(httpretty.GET,
                            "https://api.linkedin.com/v1/people/~?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json",
                            body=data,
                            content_type="application/json")

        basic_profile = self.linkedin.get_basic_profile()
        data = json.loads(data)
        expect(basic_profile).to.equal(data)


    @httpretty.activate
    def test_get_companies(self):
        data = '''
            {
              "_total": 1,
              "values":  [
                 {
                  "companyType":  {
                    "code": "O",
                    "name": "Self Owned"
                  },
                  "id": 5470551,
                  "logoUrl": "https://media.licdn.com/mpr/mpr/e34...Po.png",
                  "name": "Johni Corp"
                }
              ]
            }
        '''

        httpretty.register_uri(httpretty.GET, 
                            "https://api.linkedin.com/v1/companies:(id,name,logo-url,company-type)?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json&is-company-admin=true",
                            body=data,
                            content_type="application/json")

        basic_profile = self.linkedin.get_companies()
        data = json.loads(data)
        expect(basic_profile).to.equal(data)

    @httpretty.activate
    def test_get_profile_error_403(self):
        data = '''
        {
              "errorCode": 0,
              "message": "Member XXXX does not have permission to get company 9898",
              "requestId": "YYYY",
              "status": 403,
              "timestamp": 1441470368132
            }
        '''

        httpretty.register_uri(httpretty.GET, 
                            "https://api.linkedin.com/v1/people/~?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json",
                            body=data,
                            status=403,
                            content_type="application/json")

        with self.assertRaises(PyLinkedinAPIClientError) as ex:
            self.linkedin.get_basic_profile()

        expect(str(ex.exception)).to.equal( 'Member XXXX does not have permission to get company 9898')

    @httpretty.activate
    def test_get_profile_fields(self):
        data = '''
            {
                "id": "RdfgsgfGT",
                "numConnections": 194,
                "pictureUrl": "https://media.licdn.com/mpr/mprx/0_v0z...jQx",
                "emailAddress": "johni@johni.com"
            }
        '''
        httpretty.register_uri(httpretty.GET, 
                            'https://api.linkedin.com/v1/people/~:(id,num-connections,picture-url,email-address)?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json',
                            body=data,
                            status=200,
                            content_type="application/json")

        basic_profile = self.linkedin.get_profile(['id', 'num-connections', 'picture-url', 'email-address'])
        data = json.loads(data)
        expect(basic_profile).to.equal(data)

    @httpretty.activate
    def test_get_profile_field_not_found(self):
        data = '''
            {
              "errorCode": 0,
              "message": "Unknown field {data-profile} in resource {Person}",
              "requestId": "YUUYSSII",
              "status": 400,
              "timestamp": 1444524410254
            }
        '''

        httpretty.register_uri(httpretty.GET, 
                            'https://api.linkedin.com/v1/people/~:(id,num-connections,data-profile)?oauth2_access_token=AQVaE34QblVhvPlUn-6zWh3YLgHjx...&format=json',
                            body=data,
                            status=400,
                            content_type="application/json")

        with self.assertRaises(PyLinkedinAPIClientError) as ex:
            self.linkedin.get_profile(['id', 'num-connections', 'data-profile'])

        expect(str(ex.exception)).to.equal('Unknown field {data-profile} in resource {Person}')

    @httpretty.activate
    def test_publish_profile(self):
        pass

    @httpretty.activate
    def test_publish_company(self):
        pass


examples/basic.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from PyLinkedinAPI.PyLinkedinAPI import PyLinkedinAPI
# from PyLinkedinAPI.PyLinkedinAPI import PyLinkedinAPIClientError
# from PyLinkedinAPI.PyLinkedinAPI import PyLinkedinAPIInternalServerError


def spaces(n):
    print('\n' * n)


def get_content():

    comment = input('Insert text: ')
    title = input('Insert title: ')
    description = input('Insert description: ')
    site = input('Insert URL site: ')
    image = input('Insert URL image: ')

    return comment, title, description, site, image


def get_basic_profile():
    linkedin = PyLinkedinAPI(access_token)
    profile = linkedin.get_basic_profile()
    spaces(1)
    print(profile)


def get_companies():
    linkedin = PyLinkedinAPI(access_token)
    companies = linkedin.get_companies()
    spaces(1)
    print(companies)


def get_profile():
    linkedin = PyLinkedinAPI(access_token)
    fields = input(
        'Enter fields separated by commas without spaces \nEx: id,email-address,... : ')
    companies = linkedin.get_profile(fields.split(','))
    spaces(1)
    print(companies)


def publish_text_on_profile():
    linkedin = PyLinkedinAPI(access_token)
    comment = input('Enter text: ')
    data = linkedin.publish_profile_comment(comment)
    spaces(1)
    print(data)


def publish_text_on_company():
    linkedin = PyLinkedinAPI(access_token)
    id = int(input('Enter id(integer) company: '))
    comment = input('Enter text: ')
    data = linkedin.publish_company_comment(id, comment)
    spaces(1)
    print(data)


def publish_on_profile():
    linkedin = PyLinkedinAPI(access_token)
    comment, title, description, site, image = get_content()

    data = linkedin.publish_profile(comment,
                                    title=title,
                                    description=description,
                                    submitted_url=site,
                                    submitted_image_url=image)

    spaces(1)
    print(data)


def publish_on_company():
    linkedin = PyLinkedinAPI(access_token)

    id = int(input('Enter id(integer) company: '))
    comment, title, description, site, image = get_content()

    data = linkedin.publish_company(id,
                                    comment,
                                    title=title,
                                    description=description,
                                    submitted_url=site,
                                    submitted_image_url=image)

    spaces(1)
    print(data)


access_token = input('Insert Access Token: ')


while True:
    print(
        '''
        >>> 0 - Exit
        >>> 1 - Get Basic Profile
        >>> 2 - Get Companies
        >>> 3 - Get Profile with fields
        >>> 4 - Publish only text on Profile
        >>> 5 - Publish only text on Company
        >>> 6 - Publish content on Profile
        >>> 7 - Publish content on Company
        '''
    )

    op = input('Enter option: ')

    func = {
        '1': get_basic_profile,
        '2': get_companies,
        '3': get_profile,
        '4': publish_text_on_profile,
        '5': publish_text_on_company,
        '6': publish_on_profile,
        '7': publish_on_company,
    }

    if op not in func.keys():
        exit()

    try:
        func[op]()
    except Exception as ex:
        spaces(2)
        print('Error: {}'.format(ex))


PyLinkedinAPI/PyLinkedinAPI.py
import requests
import json


class PyLinkedinAPIClientError(Exception):
    pass


class PyLinkedinAPIInternalServerError(Exception):
    pass


class PyLinkedinAPI():

    BASE_URI_API = 'https://api.linkedin.com/v1/'

    def __init__(self, access_token):
        self.access_token = access_token
        self.default_format = 'json'

    def __factory_session(self):
        sess = requests.Session()
        return sess

    def __headers_http(self):
        headers = {'Content-Type': 'application/json', 'x-li-format': 'json'}
        return headers

    def __deserialize(self, json):
        content = json.loads(json.decode('utf-8'))
        return content

    def __publish_data_comment(self, comment):
        data = {
            "comment": comment,
            "visibility": {
                "code": "anyone"
            }
        }

        return json.dumps(data)

    def __publish_data(self, comment, **kwargs):
        title = kwargs.get('title', '')
        description = kwargs.get('description', '')
        submitted_url = kwargs.get('submitted_url', '')
        submitted_image_url = kwargs.get('submitted_image_url', '')

        data = {
            "comment": comment,
            "content": {
                "title": title,
                "description": description,
                "submitted-url": submitted_url,
                "submitted-image-url": submitted_image_url,
            },
            "visibility": {
                "code": "anyone"
            }
        }

        return json.dumps(data)

    def __build_url_get_basic_profile(self):
        url = '{url}people/~?oauth2_access_token={access_token}&\
                                            format={format}'.format(
            url=self.BASE_URI_API,
            access_token=self.access_token,
            format=self.default_format)
        return url

    def __build_url_get_companies(self):
        uri = '{url}{resource}?oauth2_access_token={access_token}&format=json&\
                                                is-company-admin=true'.format(
            url=self.BASE_URI_API,
            resource="companies:(id,name,logo-url,company-type)",
            access_token=self.access_token)

        return uri

    def __build_url_publish_profile(self):
        uri = '{url}{resource}?oauth2_access_token={access_token}&\
                                                format=json'.format(
            url=self.BASE_URI_API,
            resource='people/~/shares',
            access_token=self.access_token)

        return uri

    def __build_url_publish_company(self, id):
        resource = 'companies/{}/shares'.format(id)

        uri = '{url}{resource}?oauth2_access_token={access_token}&\
                                                format=json'.format(
            url=self.BASE_URI_API,
            resource=resource,
            access_token=self.access_token)

        return uri

    def __build_url_get_profile(self, fields):

        uri = '{url}{resource}~:({fields})?oauth2_access_token={access_token}&format=json'.format(
            url=self.BASE_URI_API,
            resource='people/',
            fields=','.join(fields),
            access_token=self.access_token)

        return uri

    def __check_response_status_code(self, resp):
        status_code = resp.status_code

        if status_code in list(range(400, 599)):
            data = json.loads(resp.content.decode('utf-8'))
            message = data['message']

            if status_code in list(range(400, 499)):
                raise PyLinkedinAPIClientError(message)
            elif status_code in list(range(500, 599)):
                raise PyLinkedinAPIInternalServerError(message)

    def __execute_request_get(self, url):
        sess = self.__factory_session()
        resp = sess.get(url, headers=self.__headers_http())
        self.__check_response_status_code(resp)
        return json.loads(resp.content.decode('utf-8'))

    def __execute_request_post(self, url, data):
        sess = self.__factory_session()
        resp = sess.post(url, data=data, headers=self.__headers_http())
        self.__check_response_status_code(resp)
        return json.loads(resp.content.decode('utf-8'))

    def get_basic_profile(self):
        url = self.__build_url_get_basic_profile()
        content = self.__execute_request_get(url)
        return content

    def get_profile(self, fields):
        url = self.__build_url_get_profile(fields)
        content = self.__execute_request_get(url)
        return content

    def get_companies(self):
        url = self.__build_url_get_companies()
        content = self.__execute_request_get(url)
        return content

    def get_detail_company(self, id):
        # https://api.linkedin.com/v1/companies/5470551:(id,name,ticker,description)?format=json
        pass

    def publish_profile(self, comment, **kwargs):
        url = self.__build_url_publish_profile()
        data = self.__publish_data(comment, **kwargs)
        return self.__execute_request_post(url, data)

    def publish_company(self, id, comment, **kwargs):
        url = self.__build_url_publish_company(id)
        data = self.__publish_data(comment, **kwargs)
        return self.__execute_request_post(url, data)

    def publish_profile_comment(self, comment):
        url = self.__build_url_publish_profile()
        data = self.__publish_data_comment(comment)
        return self.__execute_request_post(url, data)

    def publish_company_comment(self, id, comment):
        url = self.__build_url_publish_company(id)
        data = self.__publish_data_comment(comment)
        return self.__execute_request_post(url, data)
